// qase-javascript-commons/dist/client/transport/api-config-builder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveAppUrl = resolveAppUrl;
exports.createApiConfigV1 = createApiConfigV1;
exports.createApiConfigV2 = createApiConfigV2;
exports.buildHeaders = buildHeaders;
const qase_api_client_1 = require("qase-api-client");
const qase_api_v2_client_1 = require("qase-api-v2-client");
const form_data_1 = __importDefault(require("form-data"));
const DEFAULT_API_HOST = 'qase.io';
const API_BASE_URL = 'https://api-';
const APP_BASE_URL = 'https://';
const API_V1 = '/v1';
const API_V2 = '/v2';
function resolveBasePath(host, version) {
    if (host && host !== DEFAULT_API_HOST) {
        return `${API_BASE_URL}${host}${version}`;
    }
    return `https://api.${DEFAULT_API_HOST}${version}`;
}
function resolveAppUrl(config) {
    if (config.api.host && config.api.host !== DEFAULT_API_HOST) {
        return `${APP_BASE_URL}${config.api.host}`;
    }
    return `https://app.${DEFAULT_API_HOST}`;
}
function createApiConfigV1(config) {
    const apiConfig = new qase_api_client_1.Configuration({ apiKey: config.api.token, formDataCtor: form_data_1.default });
    apiConfig.basePath = resolveBasePath(config.api.host, API_V1);
    return apiConfig;
}
function createApiConfigV2(config, hostData, reporterName, frameworkName) {
    const apiConfig = new qase_api_v2_client_1.Configuration({ apiKey: config.api.token, formDataCtor: form_data_1.default });
    apiConfig.basePath = resolveBasePath(config.api.host, API_V2);
    if (hostData) {
        const headers = buildHeaders(hostData, reporterName, frameworkName);
        const existingHeaders = apiConfig.baseOptions?.headers ?? {};
        const baseOptionsWithHeaders = {
            ...(apiConfig.baseOptions ?? {}),
            headers: {
                ...existingHeaders,
                ...headers,
            },
        };
        apiConfig.baseOptions = baseOptionsWithHeaders;
    }
    return apiConfig;
}
function buildHeaders(hostData, reporterName, frameworkName) {
    const headers = {};
    const clientParts = [];
    if (reporterName?.trim()) {
        clientParts.push(`reporter=${reporterName}`);
    }
    if (hostData.reporter?.trim()) {
        clientParts.push(`reporter_version=${hostData.reporter}`);
    }
    if (frameworkName?.trim()) {
        clientParts.push(`framework=${frameworkName}`);
    }
    if (hostData.framework?.trim()) {
        clientParts.push(`framework_version=${hostData.framework}`);
    }
    if (hostData.apiClientV1?.trim()) {
        clientParts.push(`client_version_v1=${hostData.apiClientV1}`);
    }
    if (hostData.apiClientV2?.trim()) {
        clientParts.push(`client_version_v2=${hostData.apiClientV2}`);
    }
    if (hostData.commons?.trim()) {
        clientParts.push(`core_version=${hostData.commons}`);
    }
    if (clientParts.length > 0) {
        headers['X-Client'] = clientParts.join(';');
    }
    const platformParts = [];
    if (hostData.system?.trim()) {
        platformParts.push(`os=${hostData.system}`);
    }
    if (hostData.arch?.trim()) {
        platformParts.push(`arch=${hostData.arch}`);
    }
    if (hostData.language?.trim()) {
        platformParts.push(`node=${hostData.language}`);
    }
    if (hostData.packageManager?.trim()) {
        platformParts.push(`npm=${hostData.packageManager}`);
    }
    if (platformParts.length > 0) {
        headers['X-Platform'] = platformParts.join(';');
    }
    return headers;
}
