// qase-javascript-commons/dist/qase.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QaseReporter = void 0;
const chalk_1 = __importDefault(require("chalk"));
const options_1 = require("./options");
const models_1 = require("./models");
const disabled_exception_1 = require("./utils/disabled-exception");
const logger_1 = require("./utils/logger");
const hostData_1 = require("./utils/hostData");
const token_masker_1 = require("./utils/token-masker");
const state_1 = require("./state/state");
const options_resolver_1 = require("./qase/options-resolver");
const reporter_factory_1 = require("./qase/reporter-factory");
const status_processor_1 = require("./qase/status-processor");
const fallback_coordinator_1 = require("./reporters/shared/fallback-coordinator");
/**
 * @type {Record<TestStatusEnum, (test: TestResultType) => string>}
 */
const resultLogMap = {
    [models_1.TestStatusEnum.failed]: (test) => (0, chalk_1.default) `{red Test ${test.title} ${test.execution.status}}`,
    [models_1.TestStatusEnum.passed]: (test) => (0, chalk_1.default) `{green Test ${test.title} ${test.execution.status}}`,
    [models_1.TestStatusEnum.skipped]: (test) => (0, chalk_1.default) `{blueBright Test ${test.title} ${test.execution.status}}`,
    [models_1.TestStatusEnum.blocked]: (test) => (0, chalk_1.default) `{blueBright Test ${test.title} ${test.execution.status}}`,
    [models_1.TestStatusEnum.disabled]: (test) => (0, chalk_1.default) `{grey Test ${test.title} ${test.execution.status}}`,
    [models_1.TestStatusEnum.invalid]: (test) => (0, chalk_1.default) `{yellowBright Test ${test.title} ${test.execution.status}}`,
};
/**
 * @class QaseReporter
 *
 * Thin orchestrator: delegates to OptionsResolver, ReporterFactory,
 * StatusProcessor, and FallbackCoordinator. Public API is preserved.
 */
class QaseReporter {
    static instance;
    logger;
    options;
    withState;
    captureLogs;
    statusProcessor;
    fallback;
    startTestRunOperation;
    /**
     * @param {OptionsType} options
     */
    constructor(options) {
        const resolved = new options_resolver_1.OptionsResolver().resolve(options);
        this.options = resolved.composed;
        this.withState = resolved.withState;
        this.captureLogs = resolved.composed.captureLogs;
        this.logger = this.buildLogger(resolved);
        this.logger.logDebug(`Config: ${JSON.stringify((0, token_masker_1.sanitizeOptionsForLog)(resolved.composed))}`);
        const hostData = (0, hostData_1.getHostInfo)(options.frameworkPackage, options.reporterName);
        this.logger.logDebug(`Host data: ${JSON.stringify(hostData)}`);
        const factory = new reporter_factory_1.ReporterFactory(this.logger, hostData);
        const { upstream, fallback, disabled, useFallbackFromStart } = this.buildReporters(factory, resolved);
        this.statusProcessor = new status_processor_1.StatusProcessor(this.logger, resolved.composed.statusMapping, resolved.composed.testops?.statusFilter);
        this.fallback = new fallback_coordinator_1.FallbackCoordinator(this.logger, upstream, fallback, {
            onFallbackActivated: () => {
                if (this.withState) {
                    state_1.StateManager.setMode(resolved.composed.fallback);
                }
            },
            onDisabled: () => {
                if (this.withState) {
                    state_1.StateManager.setMode(options_1.ModeEnum.off);
                }
            },
        });
        if (disabled) {
            this.fallback.setDisabled(true);
        }
        if (useFallbackFromStart) {
            this.fallback.setUseFallback(true);
        }
        this.persistInitialState(resolved, disabled, useFallbackFromStart);
    }
    buildLogger(resolved) {
        const opts = { debug: resolved.composed.debug };
        if (resolved.composed.logging?.console !== undefined) {
            opts.consoleLogging = resolved.composed.logging.console;
        }
        if (resolved.composed.logging?.file !== undefined) {
            opts.fileLogging = resolved.composed.logging.file;
        }
        return new logger_1.Logger(opts);
    }
    buildReporters(factory, resolved) {
        let upstream;
        let fallbackReporter;
        let disabled = false;
        let upstreamFailed = false;
        try {
            upstream = factory.create(resolved.effectiveMode, resolved.composed, this.withState);
        }
        catch (error) {
            if (error instanceof disabled_exception_1.DisabledException) {
                disabled = true;
            }
            else {
                this.logger.logError('Unable to create upstream reporter:', error);
                if (resolved.composed.fallback === undefined) {
                    disabled = true;
                    return { upstream, fallback: fallbackReporter, disabled, useFallbackFromStart: false };
                }
                upstreamFailed = true;
            }
        }
        try {
            fallbackReporter = factory.create(resolved.effectiveFallback, resolved.composed, this.withState);
        }
        catch (error) {
            if (error instanceof disabled_exception_1.DisabledException) {
                if (upstreamFailed)
                    disabled = true;
            }
            else {
                this.logger.logError('Unable to create fallback reporter:', error);
                if (upstreamFailed && upstream === undefined)
                    disabled = true;
            }
        }
        const useFallbackFromStart = upstreamFailed && fallbackReporter !== undefined;
        return { upstream, fallback: fallbackReporter, disabled, useFallbackFromStart };
    }
    persistInitialState(resolved, disabled, useFallbackFromStart) {
        if (!this.withState)
            return;
        if (state_1.StateManager.isStateExists())
            return;
        const state = {
            RunId: undefined,
            Mode: useFallbackFromStart
                ? resolved.composed.fallback
                : resolved.composed.mode,
            IsModeChanged: undefined,
        };
        if (disabled) {
            state.Mode = options_1.ModeEnum.off;
        }
        state_1.StateManager.setState(state);
    }
    static getInstance(options) {
        if (!QaseReporter.instance) {
            QaseReporter.instance = new QaseReporter(options);
        }
        return QaseReporter.instance;
    }
    getStatusMapping() {
        return this.options.statusMapping;
    }
    async uploadAttachment(attachment) {
        const result = await this.fallback.run((r) => r.uploadAttachment(attachment), 'upload attachment');
        return result ?? '';
    }
    getResults() {
        if (this.fallback.isDisabled())
            return [];
        const active = this.fallback.isUsingFallback()
            ? this.fallback.getFallback()
            : this.fallback.getUpstream();
        return active?.getTestResults() ?? [];
    }
    setTestResults(results) {
        if (this.fallback.isDisabled())
            return;
        const active = this.fallback.isUsingFallback()
            ? this.fallback.getFallback()
            : this.fallback.getUpstream();
        active?.setTestResults(results);
    }
    async addTestResult(result) {
        if (this.fallback.isDisabled())
            return;
        const processed = this.statusProcessor.process(result);
        if (!processed)
            return;
        await this.startTestRunOperation;
        this.logger.log(resultLogMap[processed.execution.status](processed));
        await this.fallback.run((r) => r.addTestResult(processed), 'add the result');
    }
    startTestRun() {
        if (this.withState)
            state_1.StateManager.clearState();
        this.fallback.reset();
        this.logger.logDebug('Starting test run');
        const runOp = this.fallback.run(async (r) => {
            await r.startTestRun();
        }, 'start test run');
        this.startTestRunOperation = runOp.then(() => undefined);
    }
    async startTestRunAsync() {
        this.startTestRun();
        await this.startTestRunOperation;
    }
    async sendResults() {
        if (this.fallback.isDisabled())
            return;
        await this.fallback.run((r) => r.sendResults(), 'send the results');
    }
    async publish() {
        if (!this.fallback.isDisabled()) {
            await this.startTestRunOperation;
            this.logger.logDebug('Publishing test run results');
            await this.fallback.run((r) => r.publish(), 'publish the run results');
        }
        if (this.withState)
            state_1.StateManager.clearState();
    }
    async complete() {
        if (this.withState)
            state_1.StateManager.clearState();
        if (this.fallback.isDisabled())
            return;
        await this.fallback.run((r) => r.complete(), 'complete the run');
    }
    isCaptureLogs() {
        return this.captureLogs ?? false;
    }
}
exports.QaseReporter = QaseReporter;
