// qase-javascript-commons/dist/profilers/http-interceptor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpInterceptor = void 0;
exports.extractRequestInfo = extractRequestInfo;
exports.headersToRecord = headersToRecord;
exports.buildRequestStep = buildRequestStep;
const node_http_1 = __importDefault(require("node:http"));
const node_https_1 = __importDefault(require("node:https"));
const uuid_1 = require("uuid");
const test_step_1 = require("../models/test-step");
const step_execution_1 = require("../models/step-execution");
/**
 * Handles the 3 `http.request()` call signatures:
 *  - (url: string, options?, callback?)
 *  - (url: URL, options?, callback?)
 *  - (options: RequestOptions, callback?)
 *
 * Returns `{ method, url }`. Default method is 'GET'.
 */
function extractRequestInfo(args) {
    const first = args[0];
    if (typeof first === 'string') {
        // Signature: (url: string, options?, callback?)
        const options = args[1];
        const method = options && typeof options === 'object' && !('emit' in options) && 'method' in options && typeof options.method === 'string'
            ? options.method
            : 'GET';
        return { method, url: first };
    }
    if (first instanceof URL) {
        // Signature: (url: URL, options?, callback?)
        const options = args[1];
        const method = options && typeof options === 'object' && !('emit' in options) && 'method' in options && typeof options.method === 'string'
            ? options.method
            : 'GET';
        return { method, url: first.toString() };
    }
    // Signature: (options: RequestOptions, callback?)
    const opts = first;
    const method = opts.method ?? 'GET';
    const protocol = opts.protocol ?? 'http:';
    const hostname = opts.hostname ?? opts.host ?? 'localhost';
    const port = opts.port !== undefined ? `:${String(opts.port)}` : '';
    const path = opts.path ?? '/';
    const url = `${protocol}//${hostname}${port}${path}`;
    return { method, url };
}
// ─── Utility: headersToRecord ─────────────────────────────────────────────────
/**
 * Converts `http.IncomingHttpHeaders` (which has `string | string[] | undefined` values)
 * to `Record<string, string>`. Multi-value headers joined with `', '`. Undefined values skipped.
 */
function headersToRecord(headers) {
    const result = {};
    for (const [key, value] of Object.entries(headers)) {
        if (value === undefined)
            continue;
        if (Array.isArray(value)) {
            result[key] = value.join(', ');
        }
        else {
            result[key] = value;
        }
    }
    return result;
}
// ─── Utility: buildRequestStep ───────────────────────────────────────────────
function buildRequestStep(params) {
    const step = new test_step_1.TestStepType(test_step_1.StepType.REQUEST);
    step.id = (0, uuid_1.v4)();
    const data = step.data;
    data.request_method = params.method;
    data.request_url = params.url;
    data.request_headers = null;
    data.request_body = null;
    data.status_code = params.statusCode;
    data.response_body = params.responseBody;
    data.response_headers = params.responseHeaders;
    step.execution.start_time = params.startTime;
    step.execution.end_time = params.endTime;
    step.execution.duration = params.endTime - params.startTime;
    step.execution.status = params.statusCode >= 400 ? step_execution_1.StepStatusEnum.failed : step_execution_1.StepStatusEnum.passed;
    return step;
}
// ─── HttpInterceptor ─────────────────────────────────────────────────────────
/**
 * HttpInterceptor monkey-patches Node.js http/https module functions to capture
 * outgoing requests as REQUEST-type steps in the AsyncLocalStorage store.
 *
 * Pattern:
 *  1. `install()` — saves originals, replaces with wrappers
 *  2. `uninstall()` — restores originals
 */
class HttpInterceptor {
    store;
    profiler;
    origHttpRequest = null;
    origHttpGet = null;
    origHttpsRequest = null;
    origHttpsGet = null;
    constructor(store, profiler) {
        this.store = store;
        this.profiler = profiler;
    }
    install() {
        this.origHttpRequest = node_http_1.default.request;
        this.origHttpGet = node_http_1.default.get;
        this.origHttpsRequest = node_https_1.default.request;
        this.origHttpsGet = node_https_1.default.get;
        const wrappedHttpRequest = this.wrapRequestFn(this.origHttpRequest);
        const wrappedHttpGet = this.wrapRequestFn(this.origHttpGet);
        const wrappedHttpsRequest = this.wrapRequestFn(this.origHttpsRequest);
        const wrappedHttpsGet = this.wrapRequestFn(this.origHttpsGet);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
        node_http_1.default.request = wrappedHttpRequest;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
        node_http_1.default.get = wrappedHttpGet;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
        node_https_1.default.request = wrappedHttpsRequest;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
        node_https_1.default.get = wrappedHttpsGet;
    }
    uninstall() {
        if (this.origHttpRequest !== null) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
            node_http_1.default.request = this.origHttpRequest;
            this.origHttpRequest = null;
        }
        if (this.origHttpGet !== null) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
            node_http_1.default.get = this.origHttpGet;
            this.origHttpGet = null;
        }
        if (this.origHttpsRequest !== null) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
            node_https_1.default.request = this.origHttpsRequest;
            this.origHttpsRequest = null;
        }
        if (this.origHttpsGet !== null) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
            node_https_1.default.get = this.origHttpsGet;
            this.origHttpsGet = null;
        }
    }
    /**
     * Creates a wrapper around a http.request-like function.
     * The wrapper:
     *  1. Calls the original function to get the ClientRequest
     *  2. Checks if there's an active ALS context — if not, returns unchanged
     *  3. Checks shouldSkip — if true, returns unchanged
     *  4. Intercepts the 'response' event via req.emit override
     *  5. Builds a TestStepType on response end and pushes to accumulator
     */
    wrapRequestFn(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    origFn) {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const self = this;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return function (...args) {
            // Call original first to get the ClientRequest
            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
            const req = origFn.apply(this, args);
            // Extract request info
            let reqInfo;
            try {
                reqInfo = extractRequestInfo(args);
            }
            catch {
                return req;
            }
            // Check if this domain should be skipped
            if (self.profiler.shouldSkip(reqInfo.url)) {
                return req;
            }
            const startTime = Date.now();
            const origEmit = req.emit.bind(req);
            // Override req.emit to intercept the 'response' event
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            req.emit = function (event, ...emitArgs) {
                if (event === 'response') {
                    const res = emitArgs[0];
                    try {
                        const chunks = [];
                        const isError = (res.statusCode ?? 0) >= 400;
                        // Register data listener BEFORE calling original emit (Pitfall 2: stream consumption race)
                        if (isError && self.profiler.trackOnFail) {
                            res.on('data', (chunk) => {
                                try {
                                    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
                                }
                                catch {
                                    // silent failure
                                }
                            });
                        }
                        res.on('end', () => {
                            try {
                                const endTime = Date.now();
                                const statusCode = res.statusCode ?? 0;
                                const captureBody = statusCode >= 400 && self.profiler.trackOnFail;
                                const responseBody = captureBody ? Buffer.concat(chunks).toString('utf8') : null;
                                const responseHeaders = statusCode >= 400 ? headersToRecord(res.headers) : null;
                                const step = buildRequestStep({
                                    method: reqInfo.method,
                                    url: reqInfo.url,
                                    statusCode,
                                    responseBody,
                                    responseHeaders,
                                    startTime,
                                    endTime,
                                });
                                // Push to accumulator — re-fetch in case context is still active
                                const acc = self.store.getStore();
                                if (acc !== undefined && acc !== null) {
                                    acc.push(step);
                                }
                                else {
                                    // No ALS context — push to fallback accumulator (for Jest/Vitest/WDIO workers)
                                    self.profiler.fallbackSteps.push(step);
                                }
                            }
                            catch {
                                // INTC-06: silent failure — do not propagate to test
                            }
                        });
                    }
                    catch {
                        // INTC-06: silent failure in setup — do not propagate
                    }
                    // CRITICAL: Call original emit LAST (after registering data listeners)
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
                    return origEmit(event, ...emitArgs);
                }
                // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
                return origEmit(event, ...emitArgs);
            };
            return req;
        };
    }
}
exports.HttpInterceptor = HttpInterceptor;
