// qase-javascript-commons/dist/profilers/network-profiler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkProfiler = void 0;
const node_async_hooks_1 = require("node:async_hooks");
const abstract_profiler_1 = require("./abstract-profiler");
const http_interceptor_1 = require("./http-interceptor");
const fetch_interceptor_1 = require("./fetch-interceptor");
const QASE_API_HOST = 'qase.io';
// Module-level ALS store — shared across all NetworkProfiler instances to
// allow nested run() calls to see the same context.
const store = new node_async_hooks_1.AsyncLocalStorage();
class NetworkProfiler extends abstract_profiler_1.AbstractProfiler {
    skipDomains;
    // Stored for Phase 13 use (controls whether to capture network steps on test failure)
    _trackOnFail;
    /**
     * Fallback accumulator for steps captured outside any run() context.
     * Used by frameworks where test bodies cannot be wrapped in run() (Jest, Vitest, WDIO).
     * Public so interceptors can push to it via ProfilerRef.
     */
    fallbackSteps = [];
    httpInterceptor = null;
    fetchInterceptor = null;
    constructor(options = {}) {
        super();
        this.skipDomains = options.skipDomains ?? [];
        this._trackOnFail = options.trackOnFail ?? true;
    }
    get trackOnFail() {
        return this._trackOnFail;
    }
    shouldSkip(url) {
        if (url.includes(QASE_API_HOST)) {
            return true;
        }
        return this.skipDomains.some((domain) => url.includes(domain));
    }
    /**
     * Install HTTP/HTTPS monkey-patches and subscribe to undici diagnostics_channel
     * so that outgoing requests inside a `run()` context are captured as REQUEST steps.
     */
    enable() {
        this.httpInterceptor = new http_interceptor_1.HttpInterceptor(store, this);
        this.httpInterceptor.install();
        this.fetchInterceptor = new fetch_interceptor_1.FetchInterceptor(store, this);
        this.fetchInterceptor.subscribe();
    }
    /**
     * No-op per v1.3 design decision (flag-based activation model).
     * Satisfies the AbstractProfiler contract.
     */
    disable() {
        // Intentional no-op
    }
    /**
     * Uninstall HTTP/HTTPS monkey-patches and unsubscribe from undici diagnostics_channel —
     * permanently removes interception.
     */
    restore() {
        this.fetchInterceptor?.unsubscribe();
        this.fetchInterceptor = null;
        this.httpInterceptor?.uninstall();
        this.httpInterceptor = null;
    }
    /**
     * Wraps `fn` in an AsyncLocalStorage context so that any HTTP requests
     * made inside `fn` are attributed to the step accumulator for this call.
     *
     * Returns `{ result, steps }` where `result` is the return value of `fn`
     * and `steps` is the array of REQUEST steps captured during execution.
     */
    async run(fn) {
        const accumulator = [];
        const result = await store.run(accumulator, fn);
        return { result, steps: accumulator };
    }
    /**
     * Returns the current ALS store when called inside a `run()` context.
     * Returns an empty array when called outside (no active context).
     */
    getSteps() {
        return store.getStore() ?? [];
    }
    /**
     * Returns a copy of steps accumulated outside any run() context (fallback accumulator).
     * Used by frameworks where test bodies cannot be wrapped in run() (Jest, Vitest, WDIO).
     */
    getAllSteps() {
        return [...this.fallbackSteps];
    }
    /**
     * Clears the fallback step accumulator. Call after collecting steps per test.
     */
    clearSteps() {
        this.fallbackSteps.length = 0;
    }
}
exports.NetworkProfiler = NetworkProfiler;
