// qase-javascript-commons/dist/formatter/report-serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportSerializer = void 0;
const models_1 = require("../models");
/**
 * Transforms internal TestResultType / TestStepType into Qase Report
 * spec-compliant JSON (RSLT-01/02, STEP-01/02/03 rules).
 *
 * Pure: no I/O, no logger, no state. Suitable for golden-testing.
 */
class ReportSerializer {
    serializeResult(result) {
        const testopsIds = this.transformTestopsIds(result.testops_id);
        const paramGroups = this.transformGroupParams(result.group_params);
        return {
            id: result.id,
            title: result.title,
            signature: result.signature,
            execution: result.execution,
            fields: result.fields,
            attachments: result.attachments.map((a) => this.serializeAttachment(a)),
            steps: result.steps.map((s) => this.serializeStep(s)),
            params: result.params,
            param_groups: paramGroups,
            testops_ids: testopsIds,
            relations: result.relations,
            muted: result.muted,
            message: result.message,
            tags: result.tags,
        };
    }
    serializeStep(step) {
        const data = this.serializeStepData(step);
        const attachments = step.attachments.map((a) => this.serializeAttachment(a));
        return {
            id: step.id,
            step_type: step.step_type,
            data,
            parent_id: step.parent_id,
            execution: { ...step.execution, attachments },
            steps: step.steps.map((s) => this.serializeStep(s)),
        };
    }
    serializeAttachment(att) {
        return {
            id: att.id,
            file_name: att.file_name,
            mime_type: att.mime_type,
            file_path: att.file_path,
        };
    }
    transformTestopsIds(testopsId) {
        if (testopsId === null)
            return null;
        return Array.isArray(testopsId)
            ? testopsId
            : [testopsId];
    }
    transformGroupParams(groupParams) {
        const keys = Object.keys(groupParams);
        return keys.length === 0 ? [] : [keys];
    }
    serializeStepData(step) {
        const data = { ...step.data };
        if (step.step_type === models_1.StepType.TEXT && 'data' in data) {
            const textData = data;
            return {
                action: textData.action,
                expected_result: textData.expected_result,
                input_data: textData.data,
            };
        }
        if (step.step_type === models_1.StepType.GHERKIN &&
            'keyword' in data &&
            'name' in data) {
            const gherkinData = data;
            return {
                action: `${gherkinData.keyword} ${gherkinData.name}`,
                expected_result: null,
                input_data: null,
            };
        }
        if (step.step_type === models_1.StepType.REQUEST && 'request_method' in data) {
            return data;
        }
        return data;
    }
}
exports.ReportSerializer = ReportSerializer;
