// qase-javascript-commons/dist/reporters/shared/fallback-coordinator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FallbackCoordinator = void 0;
/**
 * Encapsulates the upstream → fallback → disabled cascade used across all
 * `QaseReporter` lifecycle methods. Keeps both reporters' state (useFallback
 * / disabled) in one place instead of spreading try/catch blocks everywhere.
 */
class FallbackCoordinator {
    logger;
    upstream;
    fallback;
    callbacks;
    useFallback = false;
    disabled = false;
    onDisabledFired = false;
    constructor(logger, upstream, fallback, callbacks = {}) {
        this.logger = logger;
        this.upstream = upstream;
        this.fallback = fallback;
        this.callbacks = callbacks;
    }
    isDisabled() {
        return this.disabled;
    }
    isUsingFallback() {
        return this.useFallback;
    }
    setDisabled(value) {
        this.disabled = value;
    }
    setUseFallback(value) {
        this.useFallback = value;
    }
    getUpstream() {
        return this.upstream;
    }
    getFallback() {
        return this.fallback;
    }
    /**
     * Clear the disabled / useFallback flags so the coordinator can be reused
     * for a fresh test run. Does not re-arm the `onDisabled` callback — it
     * only fires once per coordinator lifetime regardless of resets.
     */
    reset() {
        this.disabled = false;
        this.useFallback = false;
    }
    /**
     * Run `op` on the currently active reporter. On upstream failure, switch to
     * the fallback (copying accumulated results over once) and retry. On
     * fallback failure, disable the coordinator.
     *
     * Returns the operation's result, or `undefined` if the coordinator is
     * disabled or the operation produced no value.
     */
    async run(op, opName) {
        if (this.disabled)
            return undefined;
        if (this.useFallback) {
            return this.runOnFallback(op, opName);
        }
        if (!this.upstream) {
            // No upstream configured — route directly to fallback without
            // copying results (there are none to copy).
            if (!this.fallback)
                return undefined;
            this.useFallback = true;
            return this.runOnFallback(op, opName);
        }
        try {
            return await op(this.upstream);
        }
        catch (error) {
            this.logger.logError(`Unable to ${opName} in the upstream reporter:`, error);
            this.callbacks.onUpstreamFailure?.();
            if (!this.fallback) {
                this.disable();
                return undefined;
            }
            this.activateFallback();
            return this.runOnFallback(op, opName);
        }
    }
    disable() {
        if (this.disabled)
            return;
        this.disabled = true;
        if (!this.onDisabledFired) {
            this.onDisabledFired = true;
            this.callbacks.onDisabled?.();
        }
    }
    activateFallback() {
        if (this.useFallback)
            return;
        const results = this.upstream?.getTestResults() ?? [];
        this.fallback?.setTestResults(results);
        this.useFallback = true;
        this.callbacks.onFallbackActivated?.();
    }
    async runOnFallback(op, opName) {
        if (!this.fallback) {
            this.disable();
            return undefined;
        }
        try {
            return await op(this.fallback);
        }
        catch (error) {
            this.logger.logError(`Unable to ${opName} in the fallback reporter:`, error);
            this.callbacks.onFallbackFailure?.();
            this.disable();
            return undefined;
        }
    }
}
exports.FallbackCoordinator = FallbackCoordinator;
