// qase-javascript-commons/dist/utils/hostData.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getHostInfo = getHostInfo;
const os = __importStar(require("os"));
const cp = __importStar(require("child_process"));
const fs = __importStar(require("fs"));
const get_package_version_1 = require("./get-package-version");
/**
 * Normalizes platform name to unified format
 * @param {string} platform - process.platform value
 * @returns {string} Normalized platform name
 */
function normalizePlatform(platform) {
    if (platform === 'win32')
        return 'windows';
    return platform;
}
/**
 * Gets detailed OS information based on the platform
 * @returns {string} Detailed OS information
 */
function getDetailedOSInfo() {
    const platform = process.platform;
    try {
        if (platform === 'win32') {
            // Windows
            return cp.execSync('ver').toString().trim();
        }
        else if (platform === 'darwin') {
            // macOS
            return cp.execSync('sw_vers -productVersion').toString().trim();
        }
        else {
            // Linux and others
            try {
                const osRelease = fs.readFileSync('/etc/os-release', 'utf8');
                const prettyName = osRelease.match(/PRETTY_NAME="(.+)"/);
                if (prettyName?.[1]) {
                    return prettyName[1];
                }
            }
            catch {
                // Fallback if /etc/os-release doesn't exist or can't be read
            }
            return os.release();
        }
    }
    catch (error) {
        console.error('Error getting detailed OS info:', error);
        return os.release();
    }
}
/**
 * Gets npm version from environment or by executing npm command
 * @returns {string} npm version string
 */
function getNpmVersion() {
    const userAgent = process.env['npm_config_user_agent'];
    if (userAgent) {
        const match = userAgent.match(/^npm\/(\S+)/);
        if (match?.[1]) {
            return match[1];
        }
    }
    try {
        return cp.execSync('npm --version', { stdio: 'pipe' }).toString().trim();
    }
    catch {
        return '';
    }
}
/**
 * Gets information about the current host environment
 * @param {string} framework The framework name to check version for
 * @param {string} reporterName The reporter name to check version for
 * @returns {HostData} Host information object
 */
function getHostInfo(framework, reporterName) {
    try {
        return {
            system: normalizePlatform(process.platform),
            machineName: os.hostname(),
            release: os.release(),
            version: getDetailedOSInfo(),
            arch: os.arch(),
            language: process.version,
            packageManager: getNpmVersion(),
            framework: (0, get_package_version_1.getPackageVersion)(framework) ?? '',
            reporter: (0, get_package_version_1.getPackageVersion)(reporterName) ?? '',
            commons: (0, get_package_version_1.getPackageVersion)('qase-javascript-commons') ?? '',
            apiClientV1: (0, get_package_version_1.getPackageVersion)('qase-api-client') ?? '',
            apiClientV2: (0, get_package_version_1.getPackageVersion)('qase-api-v2-client') ?? '',
        };
    }
    catch (error) {
        return {
            system: normalizePlatform(process.platform),
            machineName: os.hostname() || '',
            release: os.release(),
            version: '',
            arch: os.arch(),
            language: '',
            packageManager: '',
            framework: '',
            reporter: '',
            commons: '',
            apiClientV1: '',
            apiClientV2: '',
        };
    }
}
