// qase-javascript-commons/dist/client/clientV1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientV1 = void 0;
const qase_api_client_1 = require("qase-api-client");
const api_config_builder_1 = require("./transport/api-config-builder");
const attachment_service_1 = require("./services/attachment-service");
const configuration_service_1 = require("./services/configuration-service");
const run_service_1 = require("./services/run-service");
class ClientV1 {
    logger;
    config;
    environment;
    attachmentService;
    runService;
    constructor(logger, config, environment) {
        this.logger = logger;
        this.config = config;
        this.environment = environment;
        const apiConfig = (0, api_config_builder_1.createApiConfigV1)(config);
        const appUrl = (0, api_config_builder_1.resolveAppUrl)(config);
        const configurationService = new configuration_service_1.ConfigurationService(logger, new qase_api_client_1.ConfigurationsApi(apiConfig));
        this.runService = new run_service_1.RunService(logger, new qase_api_client_1.RunsApi(apiConfig), new qase_api_client_1.EnvironmentsApi(apiConfig), configurationService, appUrl);
        this.attachmentService = new attachment_service_1.AttachmentService(logger, new qase_api_client_1.AttachmentsApi(apiConfig));
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    uploadResults(_runId, _results) {
        throw new Error('Use ClientV2 to upload results');
    }
    async createRun() {
        return this.runService.createRun(this.config, this.environment);
    }
    async completeRun(runId) {
        return this.runService.completeRun(runId, this.config);
    }
    async enablePublicReport(runId) {
        return this.runService.enablePublicReport(this.config.project, runId);
    }
    async uploadAttachment(attachment) {
        return this.attachmentService.uploadAttachment(this.config.project, attachment);
    }
}
exports.ClientV1 = ClientV1;
