// playwright-qase-reporter/dist/fixture.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.test = exports.PROFILER_CONTENT_TYPE = exports.PROFILER_ATTACHMENT_NAME = void 0;
const test_1 = require("@playwright/test");
const qase_javascript_commons_1 = require("qase-javascript-commons");
const profilers_1 = require("qase-javascript-commons/profilers");
exports.PROFILER_ATTACHMENT_NAME = 'qase-profiler-steps.json';
exports.PROFILER_CONTENT_TYPE = 'application/qase.profiler-steps+json';
// Extend the base test with a qaseProfiler fixture that:
// 1. Reads config to check if network profiling is enabled
// 2. Uses snapshot/delta pattern on fallback accumulator (diagnostics_channel
//    handlers do not inherit AsyncLocalStorage context)
// 3. Serializes captured steps as a JSON attachment
exports.test = test_1.test.extend({
    // eslint-disable-next-line no-empty-pattern
    qaseProfiler: [async ({}, use, testInfo) => {
            const configLoader = new qase_javascript_commons_1.ConfigLoader();
            const config = configLoader.load();
            const options = (0, qase_javascript_commons_1.composeOptions)({}, config);
            if (!options.profilers?.includes('network')) {
                await use();
                return;
            }
            const profiler = new profilers_1.NetworkProfiler({
                skipDomains: options.networkProfiler?.skip_domains,
                trackOnFail: options.networkProfiler?.track_on_fail,
            });
            profiler.enable();
            // Use snapshot/delta pattern with fallback accumulator because
            // diagnostics_channel handlers do not inherit AsyncLocalStorage context.
            const snapshot = profiler.getAllSteps().length;
            try {
                await use();
                const allSteps = profiler.getAllSteps();
                const steps = allSteps.slice(snapshot);
                // Serialize steps as attachment for reporter to pick up
                if (steps.length > 0) {
                    await testInfo.attach(exports.PROFILER_ATTACHMENT_NAME, {
                        contentType: exports.PROFILER_CONTENT_TYPE,
                        body: Buffer.from(JSON.stringify(steps), 'utf8'),
                    });
                }
            }
            finally {
                profiler.restore();
            }
        }, { auto: true, scope: 'test' }],
});
