// playwright-qase-reporter/dist/metadata-extractor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataExtractor = void 0;
const uuid_1 = require("uuid");
const path = __importStar(require("path"));
const playwright_1 = require("./playwright");
const PROFILER_CONTENT_TYPE = 'application/qase.profiler-steps+json';
const stepAttachRegexp = /^step_attach_(body|file)_(\w{8}-\w{4}-\w{4}-\w{4}-\w{12})_/i;
/**
 * Parses Playwright `result.attachments` into a `TestCaseMetadata` object and,
 * as a side effect, routes per-step attachments into the supplied `StepIndex`.
 *
 * Recognised attachment shapes:
 *  - `qase-metadata` content type — JSON metadata payload (ids/title/fields/...).
 *  - profiler steps content type — skipped here; merged later in onTestEnd.
 *  - `step_attach_(body|file)_<uuid>_<name>` — routed to the parent step's
 *    attachment list inside `StepIndex`.
 *  - everything else — appended to the test-level attachment list.
 */
class MetadataExtractor {
    stepIndex;
    constructor(stepIndex) {
        this.stepIndex = stepIndex;
    }
    transform(testAttachments) {
        const metadata = {
            ids: [],
            title: '',
            fields: {},
            parameters: {},
            groupParams: {},
            attachments: [],
            ignore: false,
            suite: '',
            comment: '',
            tags: [],
        };
        const attachments = [];
        for (const attachment of testAttachments) {
            if (attachment.contentType === playwright_1.ReporterContentType) {
                if (attachment.body == undefined) {
                    continue;
                }
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                const message = JSON.parse(attachment.body.toString());
                if (message.title) {
                    metadata.title = message.title;
                }
                if (message.ids) {
                    metadata.ids = message.ids;
                }
                if (message.projectMapping && typeof message.projectMapping === 'object') {
                    metadata.projectMapping = message.projectMapping;
                }
                if (message.fields) {
                    metadata.fields = message.fields;
                }
                if (message.parameters) {
                    metadata.parameters = message.parameters;
                }
                if (message.ignore) {
                    metadata.ignore = message.ignore;
                }
                if (message.suite) {
                    metadata.suite = message.suite;
                }
                if (message.comment) {
                    metadata.comment = message.comment;
                }
                if (message.groupParams) {
                    metadata.groupParams = message.groupParams;
                }
                if (message.tags) {
                    metadata.tags = [...(metadata.tags ?? []), ...message.tags];
                }
                continue;
            }
            if (attachment.contentType === PROFILER_CONTENT_TYPE) {
                // Profiler steps attachment — skip adding to test attachments.
                // Will be processed separately in onTestEnd by the reporter.
                continue;
            }
            const matches = attachment.name.match(stepAttachRegexp);
            if (matches) {
                const step = this.stepIndex.findStepByTitle(attachment.name);
                if (step) {
                    this.stepIndex.removeStep(step);
                }
                let attachmentModel;
                if (attachment.name.match(/^step_attach_body_/i)) {
                    attachmentModel = {
                        content: attachment.body == undefined ? '' : attachment.body,
                        file_name: decodeURIComponent(attachment.name.substring(matches[0].length)),
                        file_path: null,
                        mime_type: attachment.contentType,
                        size: attachment.body == undefined ? 0 : Buffer.byteLength(attachment.body),
                        id: (0, uuid_1.v4)(),
                    };
                }
                else {
                    attachmentModel = {
                        content: '',
                        file_name: decodeURIComponent(attachment.name.substring(matches[0].length)),
                        file_path: attachment.body != undefined ? attachment.body.toString() : null,
                        mime_type: attachment.contentType,
                        size: 0,
                        id: (0, uuid_1.v4)(),
                    };
                }
                if (step?.parent) {
                    this.stepIndex.addAttachmentToStep(step.parent, attachmentModel);
                    continue;
                }
                attachments.push(attachmentModel);
                continue;
            }
            const attachmentModel = {
                content: attachment.body == undefined ? '' : attachment.body,
                file_name: attachment.path == undefined ? attachment.name : path.basename(attachment.path),
                file_path: attachment.path == undefined ? null : attachment.path,
                mime_type: attachment.contentType,
                size: 0,
                id: (0, uuid_1.v4)(),
            };
            attachments.push(attachmentModel);
        }
        metadata.attachments = attachments;
        return metadata;
    }
}
exports.MetadataExtractor = MetadataExtractor;
