// playwright-qase-reporter/dist/reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlaywrightQaseReporter = void 0;
const qase_javascript_commons_1 = require("qase-javascript-commons");
const step_index_1 = require("./step-index");
const annotation_extractor_1 = require("./annotation-extractor");
const step_converter_1 = require("./step-converter");
const metadata_extractor_1 = require("./metadata-extractor");
const result_builder_1 = require("./result-builder");
/**
 * @class PlaywrightQaseReporter
 * @implements Reporter
 */
class PlaywrightQaseReporter {
    /**
     * @type {Record<TestStatus, TestStatusEnum>}
     */
    static statusMap = {
        passed: qase_javascript_commons_1.TestStatusEnum.passed,
        failed: qase_javascript_commons_1.TestStatusEnum.failed,
        skipped: qase_javascript_commons_1.TestStatusEnum.skipped,
        timedOut: qase_javascript_commons_1.TestStatusEnum.failed,
        interrupted: qase_javascript_commons_1.TestStatusEnum.failed,
    };
    /**
     * @type {Map<string, number[]>}
     * @private
     */
    static qaseIds = new Map();
    stepIndex = new step_index_1.StepIndex();
    annotationExtractor = new annotation_extractor_1.AnnotationExtractor();
    stepConverter = new step_converter_1.StepConverter(this.stepIndex);
    metadataExtractor = new metadata_extractor_1.MetadataExtractor(this.stepIndex);
    resultBuilder = new result_builder_1.ResultBuilder(this.stepConverter);
    /**
     * @type {ReporterInterface}
     * @private
     */
    reporter;
    options;
    /**
     * @param {PlaywrightQaseOptionsType} options
     * @param {ConfigLoaderInterface} configLoader
     */
    constructor(options, configLoader = new qase_javascript_commons_1.ConfigLoader()) {
        const config = configLoader.load();
        const { framework, ...composedOptions } = (0, qase_javascript_commons_1.composeOptions)(options, config);
        this.options = options.framework ?? {};
        this.reporter = qase_javascript_commons_1.QaseReporter.getInstance({
            ...composedOptions,
            frameworkPackage: '@playwright/test',
            frameworkName: 'playwright',
            reporterName: 'playwright-qase-reporter',
        });
    }
    /**
     * @param {TestCase} test
     * @param _result
     * @param step
     */
    onStepBegin(test, _result, step) {
        if (step.category !== 'test.step') {
            return;
        }
        if (this.stepIndex.hasStepCached(step)) {
            return;
        }
        this.stepIndex.cacheStep(step, test);
    }
    onBegin() {
        this.reporter.startTestRun();
    }
    /**
     * @param {TestCase} test
     * @param {TestResult} result
     */
    async onTestEnd(test, result) {
        const metadata = this.metadataExtractor.transform(result.attachments);
        const annotations = {
            ids: this.annotationExtractor.extractQaseIds(test.annotations),
            projectMapping: this.annotationExtractor.extractProjectMapping(test.annotations),
            suites: this.annotationExtractor.extractSuite(test.annotations),
        };
        const testResult = this.resultBuilder.build({
            test,
            result,
            metadata,
            annotations,
            options: this.options,
            isCaptureLogs: this.reporter.isCaptureLogs(),
            qaseIdsRegistry: PlaywrightQaseReporter.qaseIds,
        });
        if (testResult) {
            await this.reporter.addTestResult(testResult);
        }
    }
    /**
     * @returns {Promise<void>}
     */
    async onEnd() {
        await this.reporter.publish();
    }
    // add this method for supporting old version of qase
    static addIds(ids, title) {
        this.qaseIds.set(title, ids);
    }
    /**
     * @param {TestStep[]} steps
     * @returns {boolean}
     */
    checkChildrenSteps(steps) {
        return this.stepConverter.hasOnlyLeafCategories(steps);
    }
}
exports.PlaywrightQaseReporter = PlaywrightQaseReporter;
