// playwright-qase-reporter/dist/result-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResultBuilder = void 0;
const qase_javascript_commons_1 = require("qase-javascript-commons");
const internal_1 = require("qase-javascript-commons/internal");
const uuid_1 = require("uuid");
const PROFILER_CONTENT_TYPE = 'application/qase.profiler-steps+json';
const logMimeType = 'text/plain';
class ResultBuilder {
    stepConverter;
    constructor(stepConverter) {
        this.stepConverter = stepConverter;
    }
    build(args) {
        const { test, result, metadata, annotations, options, isCaptureLogs, qaseIdsRegistry } = args;
        if (metadata.ignore) {
            return null;
        }
        const error = result.error ? transformError(result.errors) : null;
        let suites = annotations.suites.length > 0
            ? annotations.suites
            : (metadata.suite ? [metadata.suite] : transformSuiteTitle(test));
        let message = null;
        if (metadata.comment !== '') {
            message = metadata.comment;
        }
        if (error) {
            if (message) {
                message += '\n\n';
            }
            else {
                message = '';
            }
            message += error.message;
        }
        if (options.browser?.addAsParameter) {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
            const browser = test._projectId ?? null;
            if (browser) {
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                metadata.parameters[options.browser?.parameterName ?? 'browser'] = browser;
                suites = suites.filter((suite) => suite !== browser);
            }
        }
        if (options.markAsFlaky && result.status === 'passed' && result.retry > 0) {
            metadata.fields['is_flaky'] = 'true';
        }
        const titleParsed = (0, qase_javascript_commons_1.parseProjectMappingFromTitle)(test.title);
        const testTitle = titleParsed.cleanedTitle || (0, internal_1.removeQaseIdsFromTitle)(test.title);
        const hasMetadataProjectMapping = metadata.projectMapping != null && Object.keys(metadata.projectMapping).length > 0;
        const hasAnnotationProjectMapping = annotations.projectMapping != null && Object.keys(annotations.projectMapping).length > 0;
        const hasTitleProjectMapping = titleParsed.projectMapping != null && Object.keys(titleParsed.projectMapping).length > 0;
        const projectMapping = hasMetadataProjectMapping
            ? metadata.projectMapping ?? null
            : hasAnnotationProjectMapping
                ? annotations.projectMapping
                : hasTitleProjectMapping
                    ? titleParsed.projectMapping
                    : null;
        const hasProjectMapping = projectMapping != null && Object.keys(projectMapping).length > 0;
        let testops_id;
        let testops_project_mapping;
        if (hasProjectMapping) {
            testops_project_mapping = projectMapping;
            testops_id = null;
        }
        else if (annotations.ids.length > 0) {
            testops_id = annotations.ids.length === 1 ? annotations.ids[0] : annotations.ids;
            testops_project_mapping = null;
        }
        else if (metadata.ids.length > 0) {
            testops_id = metadata.ids.length === 1 ? metadata.ids[0] : metadata.ids;
            testops_project_mapping = null;
        }
        else if (titleParsed.legacyIds.length > 0) {
            testops_id = titleParsed.legacyIds.length === 1 ? titleParsed.legacyIds[0] : titleParsed.legacyIds;
            testops_project_mapping = null;
        }
        else {
            testops_id = qaseIdsRegistry.get(test.title) ?? null;
            testops_project_mapping = null;
        }
        let errorForStatus = null;
        if (error) {
            errorForStatus = new Error(error.message || 'Test failed');
            if (error.stacktrace) {
                errorForStatus.stack = error.stacktrace;
            }
        }
        const testStatus = (0, qase_javascript_commons_1.determineTestStatus)(errorForStatus, result.status);
        const idsForSignature = testops_id == null ? null : (Array.isArray(testops_id) ? testops_id : [testops_id]);
        const testResult = {
            attachments: metadata.attachments,
            author: null,
            execution: {
                status: testStatus,
                start_time: result.startTime.valueOf() / 1000,
                end_time: null,
                duration: result.duration,
                stacktrace: error === null
                    ? null
                    : error.stacktrace === undefined
                        ? null
                        : error.stacktrace,
                thread: process.ppid.toString() + '-' + result.parallelIndex.toString(),
            },
            fields: metadata.fields,
            id: (0, uuid_1.v4)(),
            message,
            muted: false,
            params: metadata.parameters,
            group_params: metadata.groupParams,
            tags: metadata.tags ?? [],
            relations: {
                suite: {
                    data: suites
                        .filter((suite) => suite !== test.title)
                        .map((suite) => ({ title: suite, public_id: null })),
                },
            },
            run_id: null,
            signature: (0, qase_javascript_commons_1.generateSignature)(idsForSignature, suites, metadata.parameters),
            steps: this.stepConverter.transform(result.steps, null),
            testops_id,
            testops_project_mapping,
            title: metadata.title === '' ? testTitle : metadata.title,
        };
        if (isCaptureLogs) {
            if (result.stdout.length > 0) {
                testResult.attachments.push(convertLogsToAttachments(result.stdout, 'stdout.log'));
            }
            if (result.stderr.length > 0) {
                testResult.attachments.push(convertLogsToAttachments(result.stderr, 'stderr.log'));
            }
        }
        const profilerAttachment = result.attachments.find((a) => a.contentType === PROFILER_CONTENT_TYPE);
        if (profilerAttachment?.body) {
            try {
                const profilerSteps = JSON.parse(profilerAttachment.body.toString());
                testResult.steps = [...testResult.steps, ...profilerSteps];
            }
            catch {
                // Silent failure — corrupted profiler data must not affect test results
            }
        }
        return testResult;
    }
}
exports.ResultBuilder = ResultBuilder;
function transformError(testErrors) {
    const compound = new qase_javascript_commons_1.CompoundError();
    for (const e of testErrors) {
        if (e.message == undefined)
            continue;
        compound.addMessage(e.message);
    }
    for (const e of testErrors) {
        if (e.stack == undefined)
            continue;
        compound.addStacktrace(e.stack);
    }
    return compound;
}
function transformSuiteTitle(test) {
    return test.titlePath().filter(Boolean).map((s) => s.replace(/\\/g, '/'));
}
function convertLogsToAttachments(logs, name) {
    let content = '';
    for (const line of logs) {
        content = content + line.toString();
    }
    return {
        file_name: name,
        mime_type: logMimeType,
        content,
    };
}
