// playwright-qase-reporter/dist/step-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepConverter = void 0;
const qase_javascript_commons_1 = require("qase-javascript-commons");
const internal_1 = require("qase-javascript-commons/internal");
const uuid_1 = require("uuid");
const stepAttachRegexp = /^step_attach_(body|file)_(\w{8}-\w{4}-\w{4}-\w{4}-\w{12})_/i;
const defaultSteps = ['Before Hooks', 'After Hooks', 'Worker Cleanup'];
class StepConverter {
    stepIndex;
    constructor(stepIndex) {
        this.stepIndex = stepIndex;
    }
    transform(testSteps, parentId) {
        const steps = [];
        for (const testStep of testSteps) {
            if ((testStep.category !== 'test.step' && testStep.category !== 'hook')
                || testStep.title.match(stepAttachRegexp)) {
                continue;
            }
            if (defaultSteps.includes(testStep.title) && this.hasOnlyLeafCategories(testStep.steps)) {
                continue;
            }
            const attachments = this.stepIndex.getStepAttachments(testStep);
            const stepData = (0, internal_1.extractAndCleanStep)(testStep.title);
            const id = (0, uuid_1.v4)();
            const step = {
                id,
                step_type: qase_javascript_commons_1.StepType.TEXT,
                data: {
                    action: stepData.cleanedString,
                    expected_result: stepData.expectedResult,
                    data: stepData.data,
                },
                parent_id: parentId,
                execution: {
                    status: testStep.error ? qase_javascript_commons_1.StepStatusEnum.failed : qase_javascript_commons_1.StepStatusEnum.passed,
                    start_time: testStep.startTime.valueOf() / 1000,
                    duration: testStep.duration,
                    end_time: null,
                },
                attachments: attachments ? attachments : [],
                steps: this.transform(testStep.steps, id),
            };
            steps.push(step);
        }
        return steps;
    }
    hasOnlyLeafCategories(steps) {
        if (steps.length === 0) {
            return true;
        }
        for (const step of steps) {
            if (step.category === 'test.step' || step.category === 'hook') {
                return false;
            }
        }
        return true;
    }
}
exports.StepConverter = StepConverter;
