// playwright-qase-reporter/dist/step-index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepIndex = void 0;
/**
 * Per-instance shared state for step caching (used by `onStepBegin` and
 * `MetadataExtractor`) and per-step attachment routing (populated by
 * `MetadataExtractor`, consumed by `StepConverter`).
 */
class StepIndex {
    stepCache = new Map();
    stepAttachments = new Map();
    cacheStep(step, test) {
        this.stepCache.set(step, test);
    }
    hasStepCached(step) {
        return this.stepCache.has(step);
    }
    findStepByTitle(title) {
        for (const step of this.stepCache.keys()) {
            if (step.title === title) {
                return step;
            }
        }
        return undefined;
    }
    removeStep(step) {
        this.stepCache.delete(step);
    }
    addAttachmentToStep(parent, attachment) {
        const existing = this.stepAttachments.get(parent);
        if (existing) {
            existing.push(attachment);
        }
        else {
            this.stepAttachments.set(parent, [attachment]);
        }
    }
    getStepAttachments(step) {
        return this.stepAttachments.get(step);
    }
}
exports.StepIndex = StepIndex;
