// qase-javascript-commons/dist/client/services/api-error-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiErrorCode = void 0;
exports.processError = processError;
exports.getErrorMessage = getErrorMessage;
const is_axios_error_1 = require("../../utils/is-axios-error");
const qase_error_1 = require("../../utils/qase-error");
var ApiErrorCode;
(function (ApiErrorCode) {
    ApiErrorCode[ApiErrorCode["UNAUTHORIZED"] = 401] = "UNAUTHORIZED";
    ApiErrorCode[ApiErrorCode["FORBIDDEN"] = 403] = "FORBIDDEN";
    ApiErrorCode[ApiErrorCode["NOT_FOUND"] = 404] = "NOT_FOUND";
    ApiErrorCode[ApiErrorCode["BAD_REQUEST"] = 400] = "BAD_REQUEST";
    ApiErrorCode[ApiErrorCode["UNPROCESSABLE_ENTITY"] = 422] = "UNPROCESSABLE_ENTITY";
})(ApiErrorCode || (exports.ApiErrorCode = ApiErrorCode = {}));
function processError(error, message, model) {
    if (!(0, is_axios_error_1.isAxiosError)(error)) {
        return new qase_error_1.QaseError(message, { cause: error });
    }
    const err = error;
    const errorData = err.response?.data;
    const status = err.response?.status;
    switch (status) {
        case ApiErrorCode.UNAUTHORIZED:
            return new qase_error_1.QaseError(`${message}: Unauthorized. Please check your API token.`);
        case ApiErrorCode.FORBIDDEN:
            return new qase_error_1.QaseError(`${message}: ${errorData?.errorMessage ?? 'Forbidden'}`);
        case ApiErrorCode.NOT_FOUND:
            return new qase_error_1.QaseError(`${message}: Not found.`);
        case ApiErrorCode.BAD_REQUEST:
        case ApiErrorCode.UNPROCESSABLE_ENTITY:
            return new qase_error_1.QaseError(`${message}: Bad request\n${JSON.stringify(errorData)}\nBody: ${JSON.stringify(model)}`);
        default:
            return new qase_error_1.QaseError(message, { cause: err });
    }
}
function getErrorMessage(error) {
    if ((0, is_axios_error_1.isAxiosError)(error)) {
        const err = error;
        const errorData = err.response?.data;
        return errorData?.errorMessage ?? errorData?.error ?? errorData?.message ?? 'Unknown API error';
    }
    return error instanceof Error ? error.message : 'Unknown error';
}
