// qase-javascript-commons/dist/client/services/configuration-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigurationService = void 0;
const api_error_handler_1 = require("./api-error-handler");
class ConfigurationService {
    logger;
    configurationClient;
    constructor(logger, configurationClient) {
        this.logger = logger;
        this.configurationClient = configurationClient;
    }
    async handleConfigurations(projectCode, configurations) {
        if (!configurations.values.length) {
            return [];
        }
        const configurationIds = [];
        try {
            const existingGroups = await this.getConfigurations(projectCode);
            for (const configValue of configurations.values) {
                const { name: groupName, value: configName } = configValue;
                const group = existingGroups.find(g => g.title === groupName);
                let groupId;
                if (group) {
                    groupId = group.id;
                    this.logger.logDebug(`Found existing configuration group: ${groupName}`);
                }
                else {
                    if (configurations.createIfNotExists) {
                        const newGroupId = await this.createConfigurationGroup(projectCode, groupName);
                        if (newGroupId) {
                            groupId = newGroupId;
                            this.logger.logDebug(`Created new configuration group: ${groupName} with ID: ${groupId}`);
                        }
                        else {
                            this.logger.logDebug(`Failed to create configuration group: ${groupName}, skipping`);
                            continue;
                        }
                    }
                    else {
                        this.logger.logDebug(`Configuration group not found: ${groupName}, skipping`);
                        continue;
                    }
                }
                if (groupId) {
                    const existingConfig = group?.configurations.find(c => c.title === configName);
                    if (!existingConfig) {
                        if (configurations.createIfNotExists) {
                            const configId = await this.createConfiguration(projectCode, configName, groupId);
                            if (configId) {
                                configurationIds.push(configId);
                            }
                            this.logger.logDebug(`Created configuration: ${configName} in group: ${groupName}`);
                        }
                        else {
                            this.logger.logDebug(`Configuration not found: ${configName} in group: ${groupName}, skipping`);
                        }
                    }
                    else {
                        if (existingConfig.id) {
                            configurationIds.push(existingConfig.id);
                        }
                        this.logger.logDebug(`Configuration already exists: ${configName} in group: ${groupName}`);
                    }
                }
            }
        }
        catch (error) {
            this.logger.logError('Error handling configurations:', error);
        }
        return configurationIds;
    }
    async getConfigurations(projectCode) {
        try {
            const { data } = await this.configurationClient.getConfigurations(projectCode);
            const entities = data.result?.entities ?? [];
            return entities.map(group => ({
                id: group.id ?? 0,
                title: group.title ?? '',
                configurations: group.configurations?.map(config => ({
                    id: config.id ?? 0,
                    title: config.title ?? '',
                })) ?? [],
            }));
        }
        catch (error) {
            throw (0, api_error_handler_1.processError)(error, 'Error getting configurations');
        }
    }
    async createConfigurationGroup(projectCode, title) {
        try {
            const group = { title };
            const { data } = await this.configurationClient.createConfigurationGroup(projectCode, group);
            return data.result?.id;
        }
        catch (error) {
            throw (0, api_error_handler_1.processError)(error, 'Error creating configuration group');
        }
    }
    async createConfiguration(projectCode, title, groupId) {
        try {
            const config = { title, group_id: groupId };
            const { data } = await this.configurationClient.createConfiguration(projectCode, config);
            return data.result?.id;
        }
        catch (error) {
            throw (0, api_error_handler_1.processError)(error, 'Error creating configuration');
        }
    }
}
exports.ConfigurationService = ConfigurationService;
