// qase-javascript-commons/dist/client/services/result-transformer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResultTransformer = void 0;
const qase_api_v2_client_1 = require("qase-api-v2-client");
const models_1 = require("../../models");
const chalk_1 = __importDefault(require("chalk"));
const statusMap = {
    [models_1.TestStatusEnum.passed]: 'passed',
    [models_1.TestStatusEnum.failed]: 'failed',
    [models_1.TestStatusEnum.skipped]: 'skipped',
    [models_1.TestStatusEnum.disabled]: 'skipped',
    [models_1.TestStatusEnum.blocked]: 'blocked',
    [models_1.TestStatusEnum.invalid]: 'invalid',
};
const stepStatusMap = {
    [models_1.StepStatusEnum.passed]: qase_api_v2_client_1.ResultStepStatus.PASSED,
    [models_1.StepStatusEnum.failed]: qase_api_v2_client_1.ResultStepStatus.FAILED,
    [models_1.StepStatusEnum.blocked]: qase_api_v2_client_1.ResultStepStatus.BLOCKED,
    [models_1.StepStatusEnum.skipped]: qase_api_v2_client_1.ResultStepStatus.SKIPPED,
};
class ResultTransformer {
    logger;
    rootSuite;
    constructor(logger, rootSuite) {
        this.logger = logger;
        this.rootSuite = rootSuite;
    }
    async transform(result, attachmentUploader) {
        const attachments = await this.uploadAttachments(result.attachments, attachmentUploader);
        if (result.preparedAttachments) {
            attachments.push(...result.preparedAttachments);
        }
        const steps = await this.transformSteps(result.steps, result.title, attachmentUploader);
        const params = this.transformParams(result.params);
        const groupParams = this.transformGroupParams(result.group_params, params);
        const relations = this.getRelation(result.relations);
        const model = {
            title: result.title,
            execution: this.getExecution(result.execution),
            testops_ids: Array.isArray(result.testops_id)
                ? result.testops_id
                : result.testops_id !== null ? [result.testops_id] : null,
            attachments: attachments,
            steps: steps,
            params: params,
            param_groups: groupParams,
            relations: relations,
            message: result.message,
            fields: result.fields,
            defect: false,
            signature: result.signature,
        };
        if (result.tags && result.tags.length > 0) {
            model.fields = {
                ...model.fields,
                tags: [...new Set(result.tags)].join(','),
            };
        }
        this.logger.logDebug(`Transformed result: ${JSON.stringify(model)}`);
        return model;
    }
    async transformWithDefect(result, attachmentUploader, defect) {
        const model = await this.transform(result, attachmentUploader);
        model.defect = defect;
        return model;
    }
    async uploadAttachments(attachments, uploader) {
        const hashes = [];
        for (const attachment of attachments) {
            const hash = await uploader(attachment);
            if (hash)
                hashes.push(hash);
        }
        return hashes;
    }
    async transformSteps(steps, testTitle, attachmentUploader) {
        return Promise.all(steps.map(step => this.transformStep(step, testTitle, attachmentUploader)));
    }
    async transformStep(step, testTitle, attachmentUploader) {
        const attachmentHashes = await this.uploadAttachments(step.attachments, attachmentUploader);
        const resultStep = this.createBaseResultStep(attachmentHashes, step.execution.status);
        if (step.step_type === models_1.StepType.TEXT) {
            this.processTextStep(step, resultStep, testTitle);
        }
        else if (step.step_type === models_1.StepType.GHERKIN) {
            this.processGherkinStep(step, resultStep);
        }
        else if (step.step_type === models_1.StepType.REQUEST) {
            this.processRequestStep(step, resultStep);
        }
        if (step.steps.length > 0) {
            resultStep.steps = await this.transformSteps(step.steps, testTitle, attachmentUploader);
        }
        return resultStep;
    }
    createBaseResultStep(attachmentHashes, status) {
        return {
            data: { action: '' },
            execution: {
                status: stepStatusMap[status],
                attachments: attachmentHashes,
            },
        };
    }
    processTextStep(step, resultStep, testTitle) {
        if (!('action' in step.data) || !resultStep.data)
            return;
        const stepData = step.data;
        resultStep.data.action = stepData.action || 'Unnamed step';
        if (stepData.action === '') {
            this.logger.log((0, chalk_1.default) `{magenta Test '${testTitle}' has empty action in step. The reporter will mark this step as unnamed step.}`);
        }
        if (stepData.expected_result != null) {
            resultStep.data.expected_result = stepData.expected_result;
        }
        if (stepData.data != null) {
            resultStep.data.input_data = stepData.data;
        }
    }
    processGherkinStep(step, resultStep) {
        if (!('keyword' in step.data) || !resultStep.data)
            return;
        resultStep.data.action = step.data.keyword;
    }
    processRequestStep(step, resultStep) {
        if (!('request_method' in step.data) || !resultStep.data)
            return;
        const stepData = step.data;
        resultStep.data.action = `${stepData.request_method} ${stepData.request_url}`;
    }
    getExecution(exec) {
        return {
            status: statusMap[exec.status],
            start_time: exec.start_time,
            end_time: exec.end_time,
            duration: exec.duration,
            stacktrace: exec.stacktrace,
            thread: exec.thread,
        };
    }
    transformParams(params) {
        const transformedParams = {};
        for (const [key, value] of Object.entries(params)) {
            if (value != null) {
                transformedParams[key] = String(value);
            }
        }
        return transformedParams;
    }
    transformGroupParams(groupParams, params) {
        const keys = Object.keys(groupParams);
        if (keys.length === 0)
            return [];
        for (const [key, value] of Object.entries(groupParams)) {
            if (value) {
                params[key] = value;
            }
        }
        return [keys];
    }
    getRelation(relation) {
        if (!relation?.suite) {
            return this.getDefaultSuiteRelation();
        }
        const suiteData = this.buildSuiteData(relation.suite.data);
        return { suite: { data: suiteData } };
    }
    getDefaultSuiteRelation() {
        if (!this.rootSuite)
            return {};
        return {
            suite: {
                data: [{ public_id: null, title: this.rootSuite }],
            },
        };
    }
    buildSuiteData(suiteData) {
        const result = [];
        if (this.rootSuite) {
            result.push({ public_id: null, title: this.rootSuite });
        }
        return result.concat(suiteData.map(data => ({ public_id: null, title: data.title })));
    }
}
exports.ResultTransformer = ResultTransformer;
