// qase-javascript-commons/dist/client/services/run-service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RunService = void 0;
const qase_api_client_1 = require("qase-api-client");
const qase_error_1 = require("../../utils/qase-error");
const dateUtils_1 = require("../dateUtils");
const api_error_handler_1 = require("./api-error-handler");
const chalk_1 = __importDefault(require("chalk"));
class RunService {
    logger;
    runClient;
    environmentClient;
    configurationService;
    appUrl;
    constructor(logger, runClient, environmentClient, configurationService, appUrl) {
        this.logger = logger;
        this.runClient = runClient;
        this.environmentClient = environmentClient;
        this.configurationService = configurationService;
        this.appUrl = appUrl;
    }
    async createRun(config, environment) {
        if (config.run.id) {
            return config.run.id;
        }
        try {
            let configurationIds = [];
            if (config.configurations) {
                configurationIds = await this.configurationService.handleConfigurations(config.project, config.configurations);
            }
            const environmentId = await this.getEnvironmentId(config.project, environment);
            const runObject = this.prepareRunObject(config, environmentId, configurationIds);
            this.logger.logDebug(`Creating test run: ${JSON.stringify(runObject)}`);
            const { data } = await this.runClient.createRun(config.project, runObject);
            if (!data.result?.id) {
                throw new qase_error_1.QaseError('Failed to create test run');
            }
            this.logger.logDebug(`Test run created: ${JSON.stringify(data)}`);
            if (config.run.externalLink && data.result.id) {
                // eslint-disable-next-line @typescript-eslint/no-unsafe-enum-comparison
                const apiType = config.run.externalLink.type === 'jiraCloud'
                    ? qase_api_client_1.RunExternalIssuesTypeEnum.JIRA_CLOUD
                    : qase_api_client_1.RunExternalIssuesTypeEnum.JIRA_SERVER;
                await this.runClient.runUpdateExternalIssue(config.project, {
                    type: apiType,
                    links: [{
                            run_id: data.result.id,
                            external_issue: config.run.externalLink.link,
                        }],
                });
            }
            return data.result.id;
        }
        catch (error) {
            throw (0, api_error_handler_1.processError)(error, 'Error creating test run');
        }
    }
    async completeRun(runId, config) {
        if (!config.run.complete) {
            return;
        }
        try {
            await this.runClient.completeRun(config.project, runId);
        }
        catch (error) {
            throw (0, api_error_handler_1.processError)(error, 'Error on completing run');
        }
        if (this.appUrl) {
            const runUrl = `${this.appUrl}/run/${config.project}/dashboard/${runId}`;
            this.logger.log((0, chalk_1.default) `{blue Test run link: ${runUrl}}`);
        }
    }
    async enablePublicReport(projectCode, runId) {
        try {
            const { data } = await this.runClient.updateRunPublicity(projectCode, runId, { status: true });
            if (data.result?.url) {
                this.logger.log((0, chalk_1.default) `{blue Public report link: ${data.result.url}}`);
            }
        }
        catch (error) {
            this.logger.log((0, chalk_1.default) `{yellow Failed to generate public report link: ${(0, api_error_handler_1.getErrorMessage)(error)}}`);
        }
    }
    async getEnvironmentId(projectCode, environment) {
        if (!environment)
            return undefined;
        const { data } = await this.environmentClient.getEnvironments(projectCode, undefined, environment, 100);
        return data.result?.entities?.find((env) => env.slug === environment)?.id;
    }
    prepareRunObject(config, environmentId, configurationIds) {
        const runObject = {
            title: config.run.title ?? `Automated run ${new Date().toISOString()}`,
            description: config.run.description ?? '',
            is_autotest: true,
            cases: [],
            start_time: (0, dateUtils_1.getStartTime)(),
            tags: config.run.tags ?? [],
        };
        if (environmentId !== undefined) {
            runObject.environment_id = environmentId;
        }
        if (config.plan.id) {
            runObject.plan_id = config.plan.id;
        }
        if (configurationIds && configurationIds.length > 0) {
            runObject.configurations = configurationIds;
        }
        return runObject;
    }
}
exports.RunService = RunService;
