// qase-javascript-commons/dist/qase/options-resolver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OptionsResolver = void 0;
const env_schema_1 = __importDefault(require("env-schema"));
const env_1 = require("../env");
const options_1 = require("../options");
const state_1 = require("../state/state");
/**
 * Composes the final options used by QaseReporter:
 * - restores mode / runId from state (when withState is active)
 * - merges env-derived config with user options
 * - determines effective mode / fallback values
 */
class OptionsResolver {
    resolve(options) {
        const withState = this.detectWithState(options);
        if (withState) {
            this.restoreFromState();
        }
        const env = (0, env_1.envToConfig)((0, env_schema_1.default)({ schema: env_1.envValidationSchema }));
        const composed = (0, options_1.composeOptions)(options, env);
        return {
            effectiveMode: composed.mode ?? options_1.ModeEnum.off,
            effectiveFallback: composed.fallback ?? options_1.ModeEnum.off,
            composed,
            withState,
        };
    }
    detectWithState(options) {
        return options.frameworkName === 'cypress' || !options.frameworkName;
    }
    restoreFromState() {
        if (!state_1.StateManager.isStateExists())
            return;
        const state = state_1.StateManager.getState();
        if (state.IsModeChanged && state.Mode) {
            process.env[env_1.EnvEnum.mode] = state.Mode.toString();
        }
        if (state.RunId) {
            process.env[env_1.EnvRunEnum.id] = state.RunId.toString();
        }
    }
}
exports.OptionsResolver = OptionsResolver;
