// qase-javascript-commons/dist/qase/reporter-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReporterFactory = void 0;
const reporters_1 = require("../reporters");
const options_1 = require("../options");
const env_1 = require("../env");
const disabled_exception_1 = require("../utils/disabled-exception");
const clientV2_1 = require("../client/clientV2");
const writer_1 = require("../writer");
/**
 * Builds a mode-specific InternalReporterInterface. Throws DisabledException
 * for `ModeEnum.off` so callers can distinguish "disabled-by-config" from a
 * real failure.
 */
class ReporterFactory {
    logger;
    hostData;
    constructor(logger, hostData) {
        this.logger = logger;
        this.hostData = hostData;
    }
    create(mode, options, withState) {
        switch (mode) {
            case options_1.ModeEnum.testops:
                return this.createTestOps(options, withState);
            case options_1.ModeEnum.testops_multi:
                return this.createTestOpsMulti(options, withState);
            case options_1.ModeEnum.report:
                return this.createReport(options);
            case options_1.ModeEnum.off:
                throw new disabled_exception_1.DisabledException();
            default:
                throw new Error(`Unknown mode type`);
        }
    }
    createTestOps(options, withState) {
        if (!options.testops?.api?.token) {
            throw new Error(`Either "testops.api.token" parameter or "${env_1.EnvApiEnum.token}" environment variable is required in "testops" mode`);
        }
        if (!options.testops.project) {
            throw new Error(`Either "testops.project" parameter or "${env_1.EnvTestOpsEnum.project}" environment variable is required in "testops" mode`);
        }
        const apiClient = new clientV2_1.ClientV2(this.logger, options.testops, options.environment, options.rootSuite, this.hostData, options.reporterName, options.frameworkPackage);
        return new reporters_1.TestOpsReporter(this.logger, apiClient, withState, options.testops.project, options.testops.api.host, options.testops.batch?.size, options.testops.run?.id, options.testops.showPublicReportLink);
    }
    createTestOpsMulti(options, withState) {
        if (!options.testops?.api?.token) {
            throw new Error(`Either "testops.api.token" parameter or "${env_1.EnvApiEnum.token}" environment variable is required in "testops_multi" mode`);
        }
        const multi = options.testops_multi;
        if (!multi?.projects?.length) {
            throw new Error('"testops_multi.projects" must contain at least one project with a "code" field');
        }
        for (const p of multi.projects) {
            if (!p?.code) {
                throw new Error('Each project in "testops_multi.projects" must have a "code" field');
            }
        }
        return new reporters_1.TestOpsMultiReporter(this.logger, options.testops, multi, withState, this.hostData, options.reporterName, options.frameworkPackage, options.environment, options.testops.api?.host, options.testops.batch?.size, options.testops.showPublicReportLink);
    }
    createReport(options) {
        const localOptions = options.report?.connections?.[writer_1.DriverEnum.local];
        const writer = new writer_1.FsWriter(localOptions);
        return new reporters_1.ReportReporter(this.logger, writer, options.frameworkPackage, options.reporterName, options.environment, options.rootSuite, options.testops?.run?.id, this.hostData);
    }
}
exports.ReporterFactory = ReporterFactory;
