// qase-javascript-commons/dist/qase/status-processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatusProcessor = void 0;
const status_mapping_utils_1 = require("../utils/status-mapping-utils");
/**
 * Applies status mapping (rename rule) and status filter (drop rule) to a test
 * result. Mutates `result.execution.status` when a mapping rule matches, then
 * returns either the (possibly mutated) result or `null` if it should be
 * filtered out.
 */
class StatusProcessor {
    logger;
    statusMapping;
    statusFilter;
    constructor(logger, statusMapping, statusFilter) {
        this.logger = logger;
        this.statusMapping = statusMapping;
        this.statusFilter = statusFilter;
    }
    process(result) {
        this.applyMapping(result);
        if (this.shouldFilter(result)) {
            this.logger.logDebug(`Filtering out test result with status: ${result.execution.status}`);
            return null;
        }
        return result;
    }
    applyMapping(result) {
        if (!this.statusMapping)
            return;
        const originalStatus = result.execution.status;
        const mapped = (0, status_mapping_utils_1.applyStatusMapping)(originalStatus, this.statusMapping);
        if (mapped !== originalStatus) {
            this.logger.logDebug(`Status mapping applied: ${originalStatus} -> ${mapped}`);
            result.execution.status = mapped;
        }
    }
    shouldFilter(result) {
        if (!this.statusFilter || this.statusFilter.length === 0)
            return false;
        const statusString = result.execution.status.toString();
        this.logger.logDebug(`Checking filter: status="${statusString}", filter=${JSON.stringify(this.statusFilter)}`);
        const shouldFilter = this.statusFilter.includes(statusString);
        this.logger.logDebug(`Filter result: ${shouldFilter ? 'FILTERED' : 'NOT FILTERED'}`);
        return shouldFilter;
    }
}
exports.StatusProcessor = StatusProcessor;
